Fix/simplify alias annotation handling in op codegen.

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#32574 Fix/simplify alias annotation handling in op codegen.**
* #29986 [WIP] Some work on eager op binding codegen (gen_python_functions.py)

Previously, we ignored alias annotations when deriving argument mutability
and instead recognized particular signature patterns (in-place, out variant)
and assigned mutability accordingly. Op signatures that didn't fit these
patterns would error (e.g. see #30526, which this fixes).

No change in the generated binding code.

Code changes:
1. in function_wrapper.py, fix the mutability derivation logic used when creating an argument's c++ type property. Note that we temporarily need to trap a special case and apply the old logic, see code comment for details.

2. in gen_jit_dispatch.py, update logic that assumed only one mutable Tensor argument per declaration. Happily this mostly was accomplished by bypassing some now-redundant signature regeneration machinery. Another special case here requires that we keep the old machinery around temporarily.

Differential Revision: [D19564875](https://our.internmc.facebook.com/intern/diff/D19564875)

Commit:
Fix/simplify alias annotation handling in op codegen.

Previously, we ignored alias annotations when deriving argument mutability
and instead recognized particular signature patterns (in-place, out variant)
and assigned mutability accordingly. Op signatures that didn't fit these
patters would error (e.g. see #30526, which this fixes).

Code changes:
1. in function_wrapper.py, fix the mutability derivation logic used when
creating an argument's c++ type property. Note that we temporarily need to
trap a special case and apply the old logic, see code comment for details.

2. in gen_jit_dispatch.py, update logic that assumed only one mutable
Tensor argument per declaration. Happily this mostly was accomplished by
remove now-redundant signature regeneration machinery.

[ghstack-poisoned]

kostmo commented:
## :pill: CircleCI build failures summary and remediations
As of commit bd39bf16:


* **1/1** failures introduced in this PR


## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=bd39bf168887a36d464845f7a70554bd0d5f8350).

### :detective: 1 new failure recognized by patterns
The following build failures do not appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/4331700) pytorch_xla_linux_xenial_py3_6_clang7_build (1/1)
**Step:** "Build" ([full log](https://dr.pytorch.org/api/view-log-full?build_id=90233877) | [pattern match details](https://dr.pytorch.org/build-details.html?build_id=90233877))
<details>
<summary>
<code>Jan 24 09:30:01 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff</code>
</summary>

```
Jan 24 09:30:01 File to patch:  
Jan 24 09:30:01 Skip this patch? [y]  
Jan 24 09:30:01 4 out of 4 hunks ignored 
Jan 24 09:30:01 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-clip_grad.diff 
Jan 24 09:30:01 Applying patch file: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff 
Jan 24 09:30:01 Traceback (most recent call last): 
Jan 24 09:30:01   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 67, in <module> 
Jan 24 09:30:01     patch_repo(args) 
Jan 24 09:30:01   File "/var/lib/jenkins/workspace/xla/scripts/cond_patch.py", line 49, in patch_repo 
Jan 24 09:30:01     raise RuntimeError('Failed to apply patch: {}'.format(ppath)) 
Jan 24 09:30:01 RuntimeError: Failed to apply patch: /var/lib/jenkins/workspace/xla/torch_patches/X10-device_test.diff 
Jan 24 09:30:01 + cleanup 
Jan 24 09:30:01 + retcode=1 
Jan 24 09:30:01 + set +x 
Jan 24 09:30:01 =================== sccache compilation log =================== 
Jan 24 09:30:01 =========== If your build fails, please take a look at the log above for possible reasons =========== 
Jan 24 09:30:01 Compile requests                 1 
Jan 24 09:30:01 Compile requests executed        0 
Jan 24 09:30:01 Cache hits                       0 
Jan 24 09:30:01 Cache misses                     0 
Jan 24 09:30:01 Cache timeouts                   0 
```

</details>



---

<details><summary>This comment was automatically generated by <a href="https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment">Dr. CI</a> (expand for details).</summary>Follow <a href="https://dr.pytorch.org/admin/comments-opt-out.html">this link to opt-out</a> of these comments for your Pull Requests.

Please report bugs/suggestions on the <a href="https://github.com/kostmo/circleci-failure-tracker/issues">GitHub issue tracker</a>.
</details>

Commit:
Update on "Fix/simplify alias annotation handling in op codegen."


Previously, we ignored alias annotations when deriving argument mutability
and instead recognized particular signature patterns (in-place, out variant)
and assigned mutability accordingly. Op signatures that didn't fit these
patters would error (e.g. see #30526, which this fixes).

Code changes:
1. in function_wrapper.py, fix the mutability derivation logic used when
creating an argument's c++ type property. Note that we temporarily need to
trap a special case and apply the old logic, see code comment for details.

2. in gen_jit_dispatch.py, update logic that assumed only one mutable
Tensor argument per declaration. Happily this mostly was accomplished by
remove now-redundant signature regeneration machinery.

Codegen diff: https://github.com/bhosmer/scratch/compare/gen_jit_dispatch_before...gen_jit_dispatch_after

The difference here is due to the simplification in gen_jit_dispatch. Looking quickly at the relevant history I *think* it's a move in the right direction, but need to see test results etc. May need to follow up with another special case trap I guess

[ghstack-poisoned]

Commit:
Update on "Fix/simplify alias annotation handling in op codegen."


Previously, we ignored alias annotations when deriving argument mutability
and instead recognized particular signature patterns (in-place, out variant)
and assigned mutability accordingly. Op signatures that didn't fit these
patterns would error (e.g. see #30526, which this fixes).

No change in the generated binding code.

Code changes:
1. in function_wrapper.py, fix the mutability derivation logic used when creating an argument's c++ type property. Note that we temporarily need to trap a special case and apply the old logic, see code comment for details.

2. in gen_jit_dispatch.py, update logic that assumed only one mutable Tensor argument per declaration. Happily this mostly was accomplished by bypassing some now-redundant signature regeneration machinery. Another special case here requires that we keep the old machinery around temporarily.

[ghstack-poisoned]

Commit:
Update on "Fix/simplify alias annotation handling in op codegen."


Previously, we ignored alias annotations when deriving argument mutability
and instead recognized particular signature patterns (in-place, out variant)
and assigned mutability accordingly. Op signatures that didn't fit these
patterns would error (e.g. see #30526, which this fixes).

No change in the generated binding code.

Code changes:
1. in function_wrapper.py, fix the mutability derivation logic used when creating an argument's c++ type property. Note that we temporarily need to trap a special case and apply the old logic, see code comment for details.

2. in gen_jit_dispatch.py, update logic that assumed only one mutable Tensor argument per declaration. Happily this mostly was accomplished by bypassing some now-redundant signature regeneration machinery. Another special case here requires that we keep the old machinery around temporarily.

Differential Revision: [D19564875](https://our.internmc.facebook.com/intern/diff/D19564875)

[ghstack-poisoned]

Commit:
Update on "Fix/simplify alias annotation handling in op codegen."


Previously, we ignored alias annotations when deriving argument mutability
and instead recognized particular signature patterns (in-place, out variant)
and assigned mutability accordingly. Op signatures that didn't fit these
patterns would error (e.g. see #30526, which this fixes).

No change in the generated binding code.

Code changes:
1. in function_wrapper.py, fix the mutability derivation logic used when creating an argument's c++ type property. Note that we temporarily need to trap a special case and apply the old logic, see code comment for details.

2. in gen_jit_dispatch.py, update logic that assumed only one mutable Tensor argument per declaration. Happily this mostly was accomplished by bypassing some now-redundant signature regeneration machinery. Another special case here requires that we keep the old machinery around temporarily.

Differential Revision: [D19564875](https://our.internmc.facebook.com/intern/diff/D19564875)

[ghstack-poisoned]

Commit:
Update on "Fix/simplify alias annotation handling in op codegen."


Previously, we ignored alias annotations when deriving argument mutability
and instead recognized particular signature patterns (in-place, out variant)
and assigned mutability accordingly. Op signatures that didn't fit these
patterns would error (e.g. see #30526, which this fixes).

No change in the generated binding code.

Code changes:
1. in function_wrapper.py, fix the mutability derivation logic used when creating an argument's c++ type property. Note that we temporarily need to trap a special case and apply the old logic, see code comment for details.

2. in gen_jit_dispatch.py, update logic that assumed only one mutable Tensor argument per declaration. Happily this mostly was accomplished by bypassing some now-redundant signature regeneration machinery. Another special case here requires that we keep the old machinery around temporarily.

Differential Revision: [D19564875](https://our.internmc.facebook.com/intern/diff/D19564875)

[ghstack-poisoned]

Commit:
Update on "Fix/simplify alias annotation handling in op codegen."


Previously, we ignored alias annotations when deriving argument mutability
and instead recognized particular signature patterns (in-place, out variant)
and assigned mutability accordingly. Op signatures that didn't fit these
patterns would error (e.g. see #30526, which this fixes).

No change in the generated binding code.

Code changes:
1. in function_wrapper.py, fix the mutability derivation logic used when creating an argument's c++ type property. Note that we temporarily need to trap a special case and apply the old logic, see code comment for details.

2. in gen_jit_dispatch.py, update logic that assumed only one mutable Tensor argument per declaration. Happily this mostly was accomplished by bypassing some now-redundant signature regeneration machinery. Another special case here requires that we keep the old machinery around temporarily.

Differential Revision: [D19564875](https://our.internmc.facebook.com/intern/diff/D19564875)

[ghstack-poisoned]

Commit:
Update on "Fix/simplify alias annotation handling in op codegen."


Previously, we ignored alias annotations when deriving argument mutability
and instead recognized particular signature patterns (in-place, out variant)
and assigned mutability accordingly. Op signatures that didn't fit these
patterns would error (e.g. see #30526, which this fixes).

No change in the generated binding code.

Code changes:
1. in function_wrapper.py, fix the mutability derivation logic used when creating an argument's c++ type property. Note that we temporarily need to trap a special case and apply the old logic, see code comment for details.

2. in gen_jit_dispatch.py, update logic that assumed only one mutable Tensor argument per declaration. Happily this mostly was accomplished by bypassing some now-redundant signature regeneration machinery. Another special case here requires that we keep the old machinery around temporarily.

Differential Revision: [D19564875](https://our.internmc.facebook.com/intern/diff/D19564875)

[ghstack-poisoned]

Commit:
Update on "Fix/simplify alias annotation handling in op codegen."


Previously, we ignored alias annotations when deriving argument mutability
and instead recognized particular signature patterns (in-place, out variant)
and assigned mutability accordingly. Op signatures that didn't fit these
patterns would error (e.g. see #30526, which this fixes).

No change in the generated binding code.

Code changes:
1. in function_wrapper.py, fix the mutability derivation logic used when creating an argument's c++ type property. Note that we temporarily need to trap a special case and apply the old logic, see code comment for details.

2. in gen_jit_dispatch.py, update logic that assumed only one mutable Tensor argument per declaration. Happily this mostly was accomplished by bypassing some now-redundant signature regeneration machinery. Another special case here requires that we keep the old machinery around temporarily.

Differential Revision: [D19564875](https://our.internmc.facebook.com/intern/diff/D19564875)

[ghstack-poisoned]

zdevito reviewed (approved):
nice cleanup!

Commit:
Update on "Fix/simplify alias annotation handling in op codegen."


Previously, we ignored alias annotations when deriving argument mutability
and instead recognized particular signature patterns (in-place, out variant)
and assigned mutability accordingly. Op signatures that didn't fit these
patterns would error (e.g. see #30526, which this fixes).

No change in the generated binding code.

Code changes:
1. in function_wrapper.py, fix the mutability derivation logic used when creating an argument's c++ type property. Note that we temporarily need to trap a special case and apply the old logic, see code comment for details.

2. in gen_jit_dispatch.py, update logic that assumed only one mutable Tensor argument per declaration. Happily this mostly was accomplished by bypassing some now-redundant signature regeneration machinery. Another special case here requires that we keep the old machinery around temporarily.

Differential Revision: [D19564875](https://our.internmc.facebook.com/intern/diff/D19564875)

[ghstack-poisoned]

Commit:
Update on "Fix/simplify alias annotation handling in op codegen."


Previously, we ignored alias annotations when deriving argument mutability
and instead recognized particular signature patterns (in-place, out variant)
and assigned mutability accordingly. Op signatures that didn't fit these
patterns would error (e.g. see #30526, which this fixes).

No change in the generated binding code.

Code changes:
1. in function_wrapper.py, fix the mutability derivation logic used when creating an argument's c++ type property. Note that we temporarily need to trap a special case and apply the old logic, see code comment for details.

2. in gen_jit_dispatch.py, update logic that assumed only one mutable Tensor argument per declaration. Happily this mostly was accomplished by bypassing some now-redundant signature regeneration machinery. Another special case here requires that we keep the old machinery around temporarily.

Differential Revision: [D19564875](https://our.internmc.facebook.com/intern/diff/D19564875)

[ghstack-poisoned]

Commit:
Update on "Fix/simplify alias annotation handling in op codegen."


Previously, we ignored alias annotations when deriving argument mutability
and instead recognized particular signature patterns (in-place, out variant)
and assigned mutability accordingly. Op signatures that didn't fit these
patterns would error (e.g. see #30526, which this fixes).

No change in the generated binding code.

Code changes:
1. in function_wrapper.py, fix the mutability derivation logic used when creating an argument's c++ type property. Note that we temporarily need to trap a special case and apply the old logic, see code comment for details.

2. in gen_jit_dispatch.py, update logic that assumed only one mutable Tensor argument per declaration. Happily this mostly was accomplished by bypassing some now-redundant signature regeneration machinery. Another special case here requires that we keep the old machinery around temporarily.

Differential Revision: [D19564875](https://our.internmc.facebook.com/intern/diff/D19564875)

[ghstack-poisoned]

Commit:
Update on "Fix/simplify alias annotation handling in op codegen."


Previously, we ignored alias annotations when deriving argument mutability
and instead recognized particular signature patterns (in-place, out variant)
and assigned mutability accordingly. Op signatures that didn't fit these
patterns would error (e.g. see #30526, which this fixes).

No change in the generated binding code.

Code changes:
1. in function_wrapper.py, fix the mutability derivation logic used when creating an argument's c++ type property. Note that we temporarily need to trap a special case and apply the old logic, see code comment for details.

2. in gen_jit_dispatch.py, update logic that assumed only one mutable Tensor argument per declaration. Happily this mostly was accomplished by bypassing some now-redundant signature regeneration machinery. Another special case here requires that we keep the old machinery around temporarily.

Differential Revision: [D19564875](https://our.internmc.facebook.com/intern/diff/D19564875)

[ghstack-poisoned]
